agent/provider/src/cli/whitelist.rs

```diff
@@ -0,0 +1,200 @@
+        println_conditional(&config, "Dropped duplicated patterns:");
+        if !config.json {
+            WhitelistTable::from(DomainPatterns { patterns: skipped }).print(&config)?
+        }
+    }
+    Ok(())
+}
+
+fn remove(config: ProviderConfig, remove: Remove) -> anyhow::Result<()> {
+    let domain_patterns = DomainPatterns::load_or_create(&config.domain_whitelist_file)?;
+    let mut domain_patterns = DomainPatternIds::from(domain_patterns);
+    let (removed, _) = domain_patterns.remove(remove.ids);
+    let domain_patterns: DomainPatterns = domain_patterns.into();
+    domain_patterns.save(&config.domain_whitelist_file)?;
+    if !removed.is_empty() {
+        let table = WhitelistTable::from(DomainPatterns { patterns: removed });
+        println_conditional(&config, "Removed patterns:");
+        table.print(&config)?;
+    } else {
+        println_conditional(&config, "No matching pattern to remove.");
+        // no new pattern added, so empty list for json output
+        if config.json {
+            WhitelistTable::from(DomainPatterns {
+                patterns: Vec::new(),
+            })
+            .print(&config)?
+        }
+    };
+    Ok(())
+}
+
+struct WhitelistTable {
+    table: ResponseTable,
+}
+
+impl WhitelistTable {
+    pub fn new() -> Self {
+        let columns = vec!["ID".to_string(), "Pattern".to_string(), "Type".to_string()];
+        let values = vec![];
+        let table = ResponseTable { columns, values };
+        Self { table }
+    }
+
+    fn add(&mut self, pattern: DomainPattern) {
+        let id = pattern_to_id(&pattern);
+        let row = serde_json::json! {[ id, pattern.domain, pattern.domain_match ]};
+        self.table.values.push(row);
+    }
+
+    pub fn print(self, config: &ProviderConfig) -> anyhow::Result<()> {
+        let output = CommandOutput::from(self.table);
+        output.print(config.json)?;
+        Ok(())
+    }
+}
+
+impl From<DomainPatterns> for WhitelistTable {
+    fn from(domain_patterns: DomainPatterns) -> Self {
+        let mut table = WhitelistTable::new();
+        for pattern in domain_patterns.patterns {
+            table.add(pattern);
+        }
+        table
+    }
+}
+
+struct DomainPatternIds {
+    pattern_ids: HashMap<String, DomainPattern>,
+}
+
+impl From<DomainPatterns> for DomainPatternIds {
+    fn from(patterns: DomainPatterns) -> Self {
+        let mut pattern_ids = HashMap::new();
+        let patterns = patterns.patterns;
+        for pattern in patterns {
+            let id = pattern_to_id(&pattern);
+            pattern_ids.insert(id, pattern);
+        }
+        Self { pattern_ids }
+    }
+}
+
+impl From<DomainPatternIds> for DomainPatterns {
+    fn from(val: DomainPatternIds) -> Self {
+        let patterns = val.pattern_ids.into_values().collect();
+        DomainPatterns { patterns }
+    }
+}
+
+impl DomainPatternIds {
+    fn remove(&mut self, ids: Vec<String>) -> (Vec<DomainPattern>, Vec<String>) {
```

Comment:
I do not know what this function returns without reading its body, please add a doc comment or define a proper structure for the return type instead of a tuple.

Reply:
I really hoped you will not notice it - it was this late addition to the CLI output, and I had no patience for naming new types.
Changed it to some generic type shared between `fn remove` and `fn add`. I am aware field `skipped` gets ignored in this function, but I am implementing this CLI it without any requirements, so it is not unlikely someone will request it later. Had not done it because IMO it was unnecessary in case of `remove`.
Done.